clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -0,0 +1,191 @@
+                            subject.getPrivateCredentials().add(cred);
+                        } catch (GSSException ex) {
+                            LOG.warn("Cannot add private credential to subject; clients authentication may fail", ex);
+                        }
+                    }
+
+                    try {
+                        return Subject.doAs(subject, new PrivilegedExceptionAction<SaslServer>() {
+                                public SaslServer run() {
+                                    try {
+                                        SaslServer saslServer;
+                                        saslServer = Sasl.createSaslServer(mech, servicePrincipalName, serviceHostname, null, saslServerCallbackHandler);
+                                        return saslServer;
+                                    } catch (SaslException e) {
+                                        LOG.error("Kafka Server failed to create a SaslServer to interact with a client during session authentication: " + e);
+                                        return null;
+                                    }
+                                }
+                            });
+                    } catch (PrivilegedActionException e) {
+                        LOG.error("KafkaBroker experienced a PrivilegedActionException exception while creating a SaslServer using a JAAS principal context:" + e);
+                    }
+                } catch (IndexOutOfBoundsException e) {
+                    LOG.error("Kafka Server principal name/hostname determination error: ", e);
+                }
+            }
+        }
+        return null;
+    }
+
+    public void authenticate() throws IOException {
+        if (netOutBuffer != null && !flushNetOutBuffer()) {
+            transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+            return;
+        }
+
+        if (saslServer.isComplete()) {
+            transportLayer.addInterestOps(SelectionKey.OP_READ);
+            return;
+        }
+
+        byte[] clientToken = new byte[0];
+
+        if (netInBuffer == null) netInBuffer = new NetworkReceive(node);
+
+        long readLen = netInBuffer.readFrom(transportLayer);
+
+        if (netInBuffer.complete()) {
+            netInBuffer.payload().rewind();
+            clientToken = new byte[netInBuffer.payload().remaining()];
+            netInBuffer.payload().get(clientToken, 0, clientToken.length);
+            netInBuffer = null; // reset the networkReceive as we read all the data.
+            try {
+                byte[] response;
+                response = saslServer.evaluateResponse(clientToken);
+                if (response != null) {
+                    netOutBuffer = new NetworkSend(node, ByteBuffer.wrap(response));
+                    if (!flushNetOutBuffer()) {
+                        transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+                        return;
+                    }
+                }
```

Comment:
It seems that we need to turn off OP_WRITE after completing the send of each token. Otherwise, the server will be busy looping over the selector while waiting for the next token to be received.